File: imhere-api-gateway/src/main/java/com/kdongsu5509/imhereapigateway/filter/logger/AccessLogPrinter.java
package com.kdongsu5509.imhereapigateway.filter.logger;

import com.kdongsu5509.imhereapigateway.common.alert.port.out.MessageSendPort;
import com.kdongsu5509.imhereapigateway.domain.AccessLog;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

@Component
public class AccessLogPrinter {
    private static final Logger log = LoggerFactory.getLogger(AccessLogPrinter.class);
    private final MessageSendPort messageSendPort;
    private final AccessLogFormatter formatter;

    public AccessLogPrinter(MessageSendPort messageSendPort, AccessLogFormatter formatter) {
        this.messageSendPort = messageSendPort;
        this.formatter = formatter;
    }

    public void print(AccessLog accessLog, boolean sendAlert) {
        String formatted = formatter.format(accessLog);
        log.info(formatted);
        sendAlertIfNeeded(accessLog, formatted, sendAlert);
    }

    private void sendAlertIfNeeded(AccessLog accessLog, String formatted, boolean sendAlert) {
        if (sendAlert && accessLog.getStatus() >= 400) {
            messageSendPort.sendMessage("## 🚨 HTTP Error\n\n```json\n" + formatted + "\n```");
        }
    }
}

File: src/main/kotlin/com/kdongsu5509/support/logger/AccessLogPrinter.kt
package com.kdongsu5509.support.logger

import com.kdongsu5509.support.external.DiscordMessageDto
import com.kdongsu5509.support.external.DiscordMessageSendPort
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component

@Component
class AccessLogPrinter(
    private val discordMessageSendPort: DiscordMessageSendPort,
    private val formatter: AccessLogFormatter
) {

    @Value("\${discord.url.error:}")
    private val errorAlertChannelWebhookUrl: String? = null

    private val log = LoggerFactory.getLogger(AccessLogPrinter::class.java)

    fun print(accessLog: AccessLog, sendAlert: Boolean) {
        val formatted = formatter.format(accessLog)
        log.info(formatted)
        sendAlertIfNeeded(accessLog, formatted, sendAlert)
    }

    private fun sendAlertIfNeeded(accessLog: AccessLog, formatted: String, sendAlert: Boolean) {

        if (sendAlert && accessLog.status >= 400) {
            errorAlertChannelWebhookUrl?.let {
                discordMessageSendPort.sendMessage(
                    it,
                    DiscordMessageDto("## 🚨 HTTP Error\n\n```json\n$formatted\n```")
                )
            }
        }
    }
}
